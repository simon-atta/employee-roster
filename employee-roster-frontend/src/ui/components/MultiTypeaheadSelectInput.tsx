/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import Select from 'react-select';

export interface MultiTypeaheadSelectProps<T> {
  emptyText: string;
  options: T[];
  value: T[];
  optionToStringMap: (option: T) => string;
  onChange: (selected: T[]) => void;
  optional?: boolean;
  noClearButton?: boolean;
  autoSize?: boolean;
}

export default class MultiTypeaheadSelectInput<T> extends React.Component<MultiTypeaheadSelectProps<T>> {
  constructor(props: MultiTypeaheadSelectProps<T>) {
    super(props);
    this.onSelect = this.onSelect.bind(this);
  }

  onSelect(selections: { value: T }[]|null) {
    if (selections !== null) {
      this.props.onChange(selections.map(s => s.value));
    } else {
      this.props.onChange([]);
    }
  }

  render() {
    const { optionToStringMap, emptyText, optional, options, autoSize } = this.props;
    const selectOptions = this.props.options.map(o => ({ value: o }));
    // Why map to string first?
    // Clone of objects are not the same, and sometimes we are passed
    // clones instead of the real thing, and "===" and "==" are
    // both identity/is comparators on objects, so we compare their
    // "toString" values to check for equality
    const selected = selectOptions.filter(o => this.props.value
      .find(item => optionToStringMap(item) === optionToStringMap(o.value)) !== undefined);
    const parentStyle: React.CSSProperties = {
      position: 'relative',
    };
    if (autoSize === undefined || autoSize) {
      const maxLabelLength = Math.max(emptyText.length, ...options.map(optionToStringMap).map(s => s.length));
      parentStyle.width = `${maxLabelLength + 2}em`;
    }

    return (
      <div style={parentStyle}>
        <Select
          aria-label={emptyText}
          styles={{
            menuPortal: provided => ({
              ...provided,
              zIndex: 9999999,
            }),
          }}
          onChange={this.onSelect}
          defaultValue={selected}
          value={selected}
          selected={selected}
          placeholder={emptyText}
          getOptionLabel={o => optionToStringMap(o.value)}
          getOptionValue={o => optionToStringMap(o.value)}
          options={selectOptions}
          menuPortalTarget={document.body}
          isMulti
          isClearable={optional}
        />
      </div>
    );
  }
}
