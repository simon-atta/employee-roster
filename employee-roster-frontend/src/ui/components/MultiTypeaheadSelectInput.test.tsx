/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import MultiTypeaheadSelectInput, { MultiTypeaheadSelectProps } from './MultiTypeaheadSelectInput';

interface MockData {
  name: string;
}

type Select = MultiTypeaheadSelectInput<MockData>;
describe('MultiTypeaheadSelectInput component', () => {
  it('should remove selected from list when selected again and call onChange', () => {
    const defaultValue = [{ name: 'Option 2' }];
    const select = mount(<MultiTypeaheadSelectInput {...selectProps} value={defaultValue} />);
    (select.instance() as Select).onSelect([]);
    expect(selectProps.onChange).toBeCalled();
    expect(selectProps.onChange).toBeCalledWith([]);
  });

  it('should add option to selection when it is selected and not already in the list and call onChange', () => {
    const defaultValue = [{ name: 'Option 2' }];
    const select = mount(<MultiTypeaheadSelectInput {...selectProps} value={defaultValue} />);
    (select.instance() as Select).onSelect([...defaultValue, { name: 'Option 1' }].map(o => ({ value: o })));
    expect(selectProps.onChange).toBeCalled();
    expect(selectProps.onChange).toBeCalledWith([...defaultValue, { name: 'Option 1' }]);
  });

  it('should render correctly', () => {
    const select = shallow(<MultiTypeaheadSelectInput {...selectProps} />);
    expect(toJson(select)).toMatchSnapshot();
  });
});

const selectProps: MultiTypeaheadSelectProps<MockData> = {
  emptyText: 'Enter some data',
  options: [{ name: 'Option 1' }, { name: 'Option 2' }, { name: 'Option 3' }],
  value: [],
  optionToStringMap: jest.fn(option => option.name),
  onChange: jest.fn(),
};
