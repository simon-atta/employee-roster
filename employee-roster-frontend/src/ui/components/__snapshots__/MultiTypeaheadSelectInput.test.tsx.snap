// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiTypeaheadSelectInput component should render correctly 1`] = `
<div
  style={
    Object {
      "position": "relative",
      "width": "17em",
    }
  }
>
  <StateManager
    aria-label="Enter some data"
    defaultInputValue=""
    defaultMenuIsOpen={false}
    defaultValue={Array []}
    getOptionLabel={[Function]}
    getOptionValue={[Function]}
    isMulti={true}
    menuPortalTarget={<body />}
    onChange={[Function]}
    options={
      Array [
        Object {
          "value": Object {
            "name": "Option 1",
          },
        },
        Object {
          "value": Object {
            "name": "Option 2",
          },
        },
        Object {
          "value": Object {
            "name": "Option 3",
          },
        },
      ]
    }
    placeholder="Enter some data"
    selected={Array []}
    styles={
      Object {
        "menuPortal": [Function],
      }
    }
    value={Array []}
  />
</div>
`;
